src/Halogen/Infinite/Scroll.purs

```diff
@@ -214,61 +230,119 @@ managePreload feed = do
                                             (Map.findMin st.preloaded)
                          })
 
-manageFeed :: forall e o m .
-              Feed e m
-           => Element -> H.HalogenM (FeedState e) (FeedAction e) (FeedSlots e) o m Unit
-manageFeed feed = mask do
-  manageAbove
-  manageBelow
-  managePreload feed
-  where
-    manageBelow = do
-      state <- H.get
-      let occluded (Tuple _ page) = page.visibleHeight == 0.0 
-          extra = A.length $ A.takeWhile occluded $ A.reverse $ Map.toUnfoldable state.pages
-          target =  state.feedParams.hiddenPages + state.feedParams.preloadedPages
-          tooFew = extra < target
-          tooMany = extra > target
-      when tooFew do
-        a <- join <$> traverse (\kv -> loadPageBelow feed kv.value) (Map.findMax state.pages)
-        traverse_ (\p -> H.modify_ (\st -> st { pages = Map.insert p.id p st.pages } )) a
-      when tooMany do
-        H.modify_ (\st -> st { pages = maybe st.pages
-                                             (\kv -> Map.delete kv.key st.pages)
-                                             (Map.findMax st.pages)
-                             })
-     
-    manageAbove = do
-      st <- H.get
-      let occluded (Tuple _ page) = page.visibleHeight == 0.0 
-          extra = A.length $ A.takeWhile occluded $ Map.toUnfoldable st.pages
-          tooFew = extra < st.feedParams.hiddenPages
-          tooMany = extra > st.feedParams.hiddenPages
-      when tooFew do
-        traverse_ loadTop (Map.findMax st.preloaded) 
-      when tooMany do
-        traverse_ unloadtop (Map.findMin st.pages)
-        where
-
-          loadTop page = do
-            t <- H.liftEffect $ scrollTop feed
-            H.modify_ (\st -> st { 
-                         pages = Map.insert page.key page.value st.pages 
-                       , preloaded = Map.delete page.key st.preloaded
-                       } )
-            H.liftEffect $ setScrollTop (t + page.value.pageHeight) feed
-
-          unloadtop page = do
-            t <- H.liftEffect $ scrollTop feed
-            H.modify_ (\st -> st { 
-                         pages = Map.delete page.key st.pages
-                       , preloaded = Map.insert page.key page.value st.preloaded
-                       } )
-            H.liftEffect $ setScrollTop (t - page.value.pageHeight) feed
+data UpdateF e a =
```

Comment:
This Free monad is quite verbose but it gives some nice separation between deciding what should be done and applying changes.